npm/podController.go

```diff
@@ -472,47 +472,36 @@ func (c *podController) syncAddAndUpdatePod(newPodObj *corev1.Pod) error {
 		return nil
 	}
 
-	cachedPobObjIpChanged := false
-	// compare cached NPMPod's IP address against newPod's IP address
-	if cachedNpmPodObj.PodIP != newPodObj.Status.PodIP {
-		cachedPobObjIpChanged = true
-	}
-
 	// Dealing with "updatePod" event - Compare last applied states against current Pod states
 	// There are two possiblities for npmPodObj and newPodObj
 	// #1 case The same object with the same UID and the same key (namespace + name)
 	// #2 case Different objects with different UID, but the same key (namespace + name) due to missing some events for the old object
-	// Start processing three update cases - "ip address", label", and "portlist" change by comparing last applied states against current Pod states
 
-	deleteFromIPSets := []string{}
-	addToIPSets := []string{}
 	// compare cached NPMPod's IP address against newPod's IP address
-	if cachedPobObjIpChanged {
-		metrics.SendErrorLogAndMetric(util.PodID, "[syncAddAndUpdatePod] Info: Unexpected state. Pod (Namespace:%s, Name:%s, uid:%s, has cachedPodIp:%s which is different from PodIp:%s",
-			newPodObjNs, newPodObj.Name, cachedNpmPodObj.PodUID, cachedNpmPodObj.PodIP, newPodObj.Status.PodIP)
-		// cached PodIP needs to be cleaned up from all the cached labels
-		deleteFromIPSets = util.GetIPSetListFromLabels(cachedNpmPodObj.Labels)
-
-		// Assume that the pod IP will be released when pod moves to succeeded or failed state.
-		// If the pod transitions back to an active state, then add operation will re-establish the updated pod info.
-		// new PodIP needs to be added to all newLabels
-		addToIPSets = util.GetIPSetListFromLabels(newPodObj.Labels)
-
-		klog.Infof("Deleting pod %s %s from ipset %s and adding pod %s to ipset %s",
-			cachedNpmPodObj.PodUID, cachedNpmPodObj.PodIP, cachedNpmPodObj.Namespace, newPodObj.Status.PodIP, newPodObjNs,
-		)
-
-		// Delete the pod from its namespace's ipset.
-		if err = ipsMgr.DeleteFromSet(util.GetNSNameWithPrefix(cachedNpmPodObj.Namespace), cachedNpmPodObj.PodIP, podKey); err != nil {
-			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to delete pod from namespace ipset with err: %v", err)
+	// NPM should clean up existing references of cached pod obj and its IP.
+	// then, re-add new pod obj.
+	if cachedNpmPodObj.PodIP != newPodObj.Status.PodIP {
+		metrics.SendErrorLogAndMetric(util.PodID, "[syncAddAndUpdatePod] Info: Unexpected state. Pod (Namespace:%s, Name:%s, uid:%s, newUid:%s) , has cachedPodIp:%s which is different from PodIp:%s",
+			newPodObjNs, newPodObj.Name, cachedNpmPodObj.PodUID, string(newPodObj.UID), cachedNpmPodObj.PodIP, newPodObj.Status.PodIP)
+
+		klog.Infof("Deleting cached Pod with key:%s first due to IP Mistmatch", podKey)
+		if err = c.cleanUpDeletedPod(cachedNpmPodObj.getPodObjFromNpmPodObj()); err != nil {
```

Comment:
Would like to take the same approach from namespace controller (https://github.com/Azure/azure-container-networking/blob/2c74675e182a1608b6ef606a8e171d25d7e91813/npm/nameSpaceController.go#L438) - use "cachedNpmPodKey" as parameter for cleanUpDeletedPod function, not corev1.Pod object. So, we can remove unnecessary creation of PodObj  and getPodObjFromNpmPodObj function.

Reply:
Done makes sense